orc8r/cloud/go/obsidian/access/rest_middleware.go

```diff
@@ -100,3 +104,104 @@ func makeErr(decorate logDecorator, status int, errFmt string, errArgs ...interf
 	glog.V(1).Infof("REST middleware (obsidian) rejected request: %s", decorate(errFmt, errArgs...))
 	return echo.NewHTTPError(status, fmt.Sprintf(errFmt, errArgs...))
 }
+
+// TokenMiddleware parses the <username>:<token> from the header, validates the token,
+// then checks if the request is within the specified permissions granted to the user
+func TokenMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		req := c.Request()
+		header := req.Header.Get(CLIENT_ACCESS_TOKEN_KEY)
+
+		username, token, err := parseAuthHeader(header)
+		if err != nil {
+			// TODO(christinewang5): remove after bootstrapping admin token
+			glog.Error("no header found, bootstrapping admin user")
+			adminToken, err := bootstrapAdmin(req.Context())
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, err)
+			}
+			username = RootUsername
+			token = adminToken
+			// return echo.NewHTTPError(http.StatusBadRequest, err)
+		}
+
+		if err := certifier.ValidateToken(token); err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, err)
+		}
```

Comment:
Wait does it not already checks it for every request just by being in the middleware? 
The reason I put this here and not in the service is so that we don't have to make yet another rpc call to the service, we can return early if the token isn't valid. 
But lmk if you mean sth else!

Reply:
Ah right. The idea I had is -- currently we perform a PD at the REST API middleware. But we could, and might eventually, pass this token around with each gRPC request. In which case we would have multiple callers needing a token->PD conversion

It's def not wrong to do it here, but I think it will be simpler to move it to the certifier service -- 1 owner

Reply:
Sureee thing!